### PR TITLE
prevents redundant memcached compression

### LINKED ISSUE #39748: memcached double compression, ignores compression threshold
Associated PR: https://github.com/rails/rails/pull/39747

### Steps to reproduce

Initialize `MemCacheStore` with `compress: true` and `compression_threshold` to some value greater than `1 kb` e.g. `10 kb`.

For more detailed steps to reproduce, see the [tests in the associated PR](https://github.com/rails/rails/pull/39747/files#diff-d6b6ccc47f3150ac15b66251e5661663R123).

### Expected behavior

When writing values whose size is `< 10 kb`, those values should not be compressed. 

When writing values much larger than `10 kb`, which when compressed are still larger than `10 kb`, those values should only be compressed once each.

### Actual behavior

When writing values whose size is `< 10 kb` and `> 1 kb` (the default `Dalli` `compression_min_size`), those values are compressed by `Dalli`.  

When writing values much larger than `10 kb`, which when compressed are still larger than `10 kb`, those values are compressed twice each.

### System configuration
**Rails version**: 6.1.0.alpha
**Ruby version**: 2.6.5

This bug seems to be longstanding, extending back to at least Rails 5.2.

### PR DESCRIPTION
**Pull request by maxgurewitz**
- Disables Dalli compression in MemCacheStore.
- Fixes issue where redundant compression in Dalli can cause values to
either be compressed twice, or compressed when they fall below the
specified compression threshold.

### CODE REVIEW

**Comment by casperisfine** [comment]
I just noticed that problem as well. However I think it would be better solved by disabling the compression done by ActiveSupport.

**Comment by maxgurewitz** [comment]
@casperisfine I considered that, but came to the conclusion that there'd be a bit more code involved in doing that. Any reason you prefer performing compression in `Dalli`?

**Comment by casperisfine** [comment]
> Any reason you prefer performing compression in `Dalli`?

Yes, it's because Dalli store in memcached client flags wether the value was compressed or not: https://github.com/petergoldstein/dalli/blob/ca8b286de230546c3bfe2789e9b6bc5a2c619570/lib/dalli/server.rb#L435-L436 so it's a bit more efficient.

Additionally Dalli allow you to replace the compression mechanism, so it's quite superior to `AS::Cache` in that regard.

I'm currently working on refactoring to make all that easy, but I don't know where it will get at this stage, so maybe you PR make sense short term.

File: activesupport/test/cache/stores/mem_cache_store_test.rb [hunk 1]
```diff
@@ -120,7 +120,43 @@ def test_read_should_return_a_different_object_id_each_time_it_is_called
     assert_not_equal value, @cache.read("foo")
   end
 
+  def test_no_compress_when_below_throwshold
```
Comment (p8) [inline activesupport/test/cache/stores/mem_cache_store_test.rb:123]:
```suggestion
  def test_no_compress_when_below_threshold
```

**Comment by maxgurewitz** [comment]
@p8 rebased!

File: activesupport/lib/active_support/cache/mem_cache_store.rb [hunk 2]
```diff
@@ -141,6 +141,8 @@ def read_entry(key, **options)
 
         # Write an entry to the cache.
         def write_entry(key, entry, **options)
+          # Prevents duplicate compression within Dalli.
+          options = options.merge(compress: false)
```
Comment (eugeneius) [inline activesupport/lib/active_support/cache/mem_cache_store.rb:145]: Passing `compress: false` directly as a keyword argument is a bit more efficient:

```patch
diff --git a/activesupport/lib/active_support/cache/mem_cache_store.rb b/activesupport/lib/active_support/cache/mem_cache_store.rb
index f646c6ff82..ae3102406d 100644
--- a/activesupport/lib/active_support/cache/mem_cache_store.rb
+++ b/activesupport/lib/active_support/cache/mem_cache_store.rb
@@ -149,7 +149,7 @@ def write_entry(key, entry, **options)
             expires_in += 5.minutes
           end
           rescue_error_with false do
-            @data.with { |c| c.send(method, key, value, expires_in, **options) }
+            @data.with { |c| c.send(method, key, value, expires_in, **options, compress: false) }
           end
         end
```

```bash
$ ruby -r benchmark-memory -e "def foo(**) end; options = {}; Benchmark.memory { |x| x.report { options = options.merge(compress: false); foo(**options) } }"
Calculating -------------------------------------
                       672.000  memsize (   168.000  retained)
                         4.000  objects (     1.000  retained)
                         0.000  strings (     0.000  retained)
$ ruby -r benchmark-memory -e "def foo(**); end; options = {}; Benchmark.memory { |x| x.report { foo(**options, compress: false) } }"
Calculating -------------------------------------
                       336.000  memsize (     0.000  retained)
                         2.000  objects (     0.000  retained)
                         0.000  strings (     0.000  retained)
```

File: activesupport/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,8 @@
+*   Fixes issue in `ActiveSupport::Cache::MemCacheStore` which can cause duplicate compression,
+    or cause the provided `compression_threshold` to not be respected.
```
Comment (eugeneius) [inline activesupport/CHANGELOG.md:2]: The changelog entry is for a bug that's now fixed, let's describe it in the past tense.

```suggestion
*   Fix an issue in `ActiveSupport::Cache::MemCacheStore` which caused duplicate compression,
    and caused the provided `compression_threshold` to not always be respected.
```

**Review by eugeneius — COMMENTED** [review]
I'd like to merge this now even if it's obsoleted by @casperisfine's work later, to ensure this problem is fixed in 6.1.

**Comment by eugeneius** [comment]
Thanks! ⭐

**Comment by casperisfine** [comment]
> I'd like to merge this now even if it's obsoleted by @casperisfine's work later, to ensure this problem is fixed in 6.1.

Yeah that's fine. I'm blocked in early stage refactoring (https://github.com/rails/rails/pull/39770) because I can't find someone to review. It might be a bit late for 6.1 anyway, so we might as well get this quick fix out now.

**Pull request merged**